cocotb/handle.py

```diff
@@ -530,11 +519,67 @@ def _range_iter(self, left, right):
+
+        **Getting value**:
+
+        The returned list contains the values of each sub-handle
+        in the order indicated in the table:
+
+        +--------------+---------------------+--------------------------------------+
+        | Verilog      | VHDL                | Python list                          |
+        +==============+=====================+======================================+
+        | ``arr[4:7]`` | ``arr(4 to 7)``     | ``[arr[4], arr[5], arr[6], arr[7]]`` |
+        +--------------+---------------------+--------------------------------------+
+        | ``arr[7:4]`` | ``arr(7 downto 4)`` | ``[arr[7], arr[6], arr[5], arr[4]]`` |
+        +--------------+---------------------+--------------------------------------+
+
+        **Setting value**:
+
+        The indices in the list ``value`` will be assigned to sub-handles
+        in the order indicated in the table:
+
+        +--------------+---------------------+--------------------------------+
+        | Verilog      | VHDL                | Array values                   |
+        +==============+=====================+================================+
+        | ``arr[4:7]`` | ``arr(4 to 7)``     | - ``arr[4]`` gets ``value[0]`` |
+        |              |                     | - ...                          |
+        |              |                     | - ``arr[7]`` gets ``value[3]`` |
+        +--------------+---------------------+--------------------------------+
+        | ``arr[7:4]`` | ``arr(7 downto 4)`` | - ``arr[7]`` gets ``value[0]`` |
+        |              |                     | - ...                          |
+        |              |                     | - ``arr[7]`` gets ``value[3]`` |
+        +--------------+---------------------+--------------------------------+
+
+        .. warning::
+            Assigning a value to a sub-handle:
+
+            - **Wrong**: ``dut.some_array.value[0] = 1`` (Gets value as a list then updates index 0)
+            - **Right**: ``dut.some_array[0].value = 1``
+        """
+        # Don't use self.__iter__, because it has an unwanted `except IndexError`
+        return [
+            self[i]
+            for i in self._range_iter(self._range[0], self._range[1])
+        ]
+
+    @value.setter
+    def value(self, value):
+        """Assign value from a list of same length to an array in left-to-right order.
+        Index 0 of the list maps to the left-most index in the array.
+
+        See getter docstring above.
+        """
+        if type(value) is not list:
+            raise TypeError("Assigning non-list value to object %s of type %s" % (self._name, type(self)))
+        if len(value) != len(self):
+            raise IndexError("Assigning list of length %d to object %s of length %d" % (
+                len(value), self._name, len(self)))
+        for val_idx, self_idx in enumerate(self._range_iter(self._range[0], self._range[1])):
+            self[self_idx] = value[val_idx]
```

Comment:
I'm thinking this should maybe be
```suggestion
    @value.setter
    def setimmediatevalue(self, value):
        """Assign value from a list of same length to an array in left-to-right order.
        Index 0 of the list maps to the left-most index in the array.

        See the docstring for :attr:`value` above
        """
        if type(value) is not list:
            raise TypeError("Assigning non-list value to object %s of type %s" % (self._name, type(self)))
        if len(value) != len(self):
            raise IndexError("Assigning list of length %d to object %s of length %d" % (
                len(value), self._name, len(self)))
        for val_idx, self_idx in enumerate(self._range_iter(self._range[0], self._range[1])):
            self[self_idx].setimmediatevalue(value[val_idx])
```

Reply:
This would break the `handle <= value` semantics of using the cached values.
`setimmediatevalue` only works for directly modifiable objects atm.

Reply:
Spotted a typo and updated.

You're right, for `setimmediatevalue` to work, this would need to move:
https://github.com/cocotb/cocotb/blob/8248bd336bdb0a641334f17a18176833428ae6fd/cocotb/handle.py#L653-L661

Since it doesn't implement `setimmediatevalue` at all today, I'm fine with putting it off till a later patch.

Reply:
I could implement both.

Reply:
Let's leave that to a follow-up, and get this in as is.